### PR TITLE
Caching in Cnfizer

### PR DESCRIPTION
**Pull request by blishko**
This PR proposes a strategy for caching terms already processed by Cnfizer to avoid duplicate clauses in the SAT solver. (issue #23 )

Caching happens on two levels: 
- in Cnfizer to avoid processing the same top-level formula that has already been asserted to the SAT solver.
- in Tseitin to remember already cnfized terms to avoid duplicate clauses defining complex terms (such as conjunction, disjunction, xor, iff, etc.)

In addition, this PR contains a small fix to correctly process QF_RDL and QF_IDL logics by the interpreter.

### CODE REVIEW

**Comment by blishko** [comment]
@aehyvari Let me know what you think and if you have any ideas for improvement. Currently the cache remembers terms for a given frame.

**Comment by blishko** [comment]
Seems like a (commented) output for one regression test has changed: ./splitting/p2-zenonumeric_s6-deep.smt2
The test cases outputs some clauses and these clauses has now changed.

@aehyvari What is the meaning of these printed clauses?

File: src/cnfizers/Cnfizer.cc [hunk 1]
```diff
@@ -212,6 +215,20 @@ lbool Cnfizer::cnfizeAndGiveToSolver(PTRef formula, FrameId frame_id)
...
+    assert(formula != PTRef_Undef);
+    // Top level formula must not be and anymore
+    assert(!logic.isAnd(formula));
+    bool res = true;
+    // Add the top level literal as a unit to solver.
+    vec<Lit> clause;
+    clause.push(this->getOrCreateLiteralFor(formula));
+    res &= addClause(clause);
+
+    res = cnfize(formula);
```
[the remark this answers is not in the document]
Comment (blishko) [reply src/cnfizers/Cnfizer.cc:228]: Yes, that makes sense.

**Comment by blishko** [comment]
> It looks good. Maybe one question: I didn't understand how the top-level definitions are handled now with respect to the definition literal? Are they produced or is that case handled somewhere else?

Not sure I understand. There are two options that Cnfizer can do we a term: cnfize or cnfizeAndAssert. Cnfize means it will just produce the clauses encoding the term, but it will assert to the solver that this term must be true.
cnfizeAndAssert adds exactly this, it will call cnfize to add the encoding clauses, and on top of that it will add the unit clause with the variable representing this term, effectively saying to the solver: this term must be true.

File: src/cnfizers/Cnfizer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (blishko) [reply src/cnfizers/Cnfizer.cc]: Fixed

**Comment by aehyvari** [comment]
Ok I think I see how it works.  Bottom line is that there is not going to be a unit clause for a top-level disjunction for instance.

**Comment by blishko** [comment]
> Ok I think I see how it works. Bottom line is that there is not going to be a unit clause for a top-level disjunction for instance.

On the contrary, there **will be** such a unit clause. Previously there was a weird (to me) distinction that if the top level formula is OR you don't insert the unit clause, but if the top level was anything else (for example an implication) you do insert it. Because of this, there was a special need to distinguish cases when processing OR which made the code more complex. I don't see a good reason why top level disjunction should be treated differently than other top level formulas.

**Pull request merged**